### ISSUE TITLE
Detokenize?

### ISSUE THREAD

**Issue by hattonjs** [Natooz/MidiTok#43]
The documentation is a little lacking.  Can you add some documentation that discusses the detokenization of a BPE token dataset?  I would like to see an example or two.

**Comment by Natooz** [comment]
Hello,

Of course this can be done. I'll add this to the backlog and do it when I'll have time.
In short:
In MidiTok, there is a *base* vocabulary composed of the basic tokens that represent note attributes and time events. These tokens can be seen as the equivalent of characters (or bytes) for natural language.
A BPE token *id* (integer, or index of the token in the vocabulary) is linked to a list of bytes, each of these bytes are linked to basic tokens, these basics tokens are used to create the generated notes / MIDI.

**Comment by hattonjs** [comment]
I figured it could be done.  However, it would be huge if there were examples on how to use what is a huge part of the functionality of this library.  It is crucial to most projects to detokenize and not all of us are at the same skill level but would like to use the library.  I particularly interested in how to generate midi from BPE encoded tokens.   Your documentation is great for the tokenizing part.  With that said … I am able to detokenize from regular tokens … there is just nothing with regards to doing that from BPE.  Also, if you could add ALL the properties and methods to the docs with a brief explanation of each … I would be huge and make this library much more usable.

THANK YOU!

On May 23, 2023, at 8:13 AM, Nathan Fradet ***@***.***> wrote:

﻿

Hello,

Of course this can be done. I'll add this to the backlog and do it when I'll have time.
In short:
In MidiTok, there is a base vocabulary composed of the basic tokens that represent note attributes and time events. These tokens can be seen as the equivalent of characters (or bytes) for natural language.
A BPE token id (integer, or index of the token in the vocabulary) is linked to a list of bytes, each of these bytes are linked to basic tokens, these basics tokens are used to create the generated notes / MIDI.

**Comment by Natooz** [comment]
Hi, 👋

FIY the [BPE doc](https://miditok.readthedocs.io/en/latest/bpe.html) has just been updated.
Don't hesitate to tell me if there are still things you find unclear or not well explained.

**Comment by hattonjs** [comment]
Thank you!  I will check it out.

Sent from my iPhone

On May 29, 2023, at 12:26 PM, Nathan Fradet ***@***.***> wrote:

﻿

Hi, 👋

FIY the BPE doc<https://miditok.readthedocs.io/en/latest/bpe.html> has just been updated.
Don't hesitate to tell me if there are still things you find unclear or not well explained.

**Comment by hattonjs** [comment]
I would like to see an example of how to remove BPE from the tokenization so I can convert back to MIDI.  Is this possible?

**Comment by Natooz** [comment]
You have it in the doc, using the `decode_bpe` method.

```Python
tokens = tokenizer.load_tokens(token_paths[0])
tokens = TokSequence(ids=tokens)
tokens_with_bpe = tokenizer.apply_bpe(deepcopy(tokens))  # copy as the method is inplace
tokens_no_bpe = tokenizer.decode_bpe(deepcopy(tokens_with_bpe))
```

If you need more details on the exact decoding procedure, I suggest you to look at the source code, and the source code of the hugging face tokenizers library (written in Rust)
